src/ILCompiler/src/Program.cs

```diff
@@ -402,15 +402,25 @@ private int Run(string[] args)
                     foreach (var inputFile in typeSystemContext.InputFilePaths)
                     {
                         EcmaModule module = typeSystemContext.GetModuleFromPath(inputFile.Value);
-
                         compilationRoots.Add(new ReadyToRunRootProvider(module));
                         inputModules.Add(module);
+
                         if (!_isInputVersionBubble)
                         {
                             break;
                         }
                     }
 
+                    if (_isInputVersionBubble)
```

Comment:
I wonder if our compilation module group needs to support more detail so we don't have to put all reference assemblies into the input modules list. It seems like it needs three settings now:
1. Assembly is going to be compiled
2. Assembly is in the bubble so we can inline versionable code
3. Assembly is not in the bubble (a traditional reference file in CoreRT)

Would it simplify the book-keeping you had to do with a global input module if we extended the compilation module group to also allow specifying how "in" the group we care about? Ie, when deciding to compile a dependent method, we'd ask `CompilationModuleGroup.ContainsType(Type, CompilationLevel.Compile)`, whereas in somewhere like `CanInline` we'd ask `CompilationModuleGroup.ContainsType(Type, CompilationLevel.InBubble)` for example. It would be nice to keep a single place we ask "is this thing in or out for this compilation" even with expanded version bubbles or compiling multiple assemblies into one ready-to-run artifact.

Reply:
I agree that my change effectively does assign several "levels" to the assembly closure. At some point I was even thinking whether the "in-bubble reference assemblies" should be in CompilationModuleGroup at all but I tend to agree with you that centralizing these kinds of queries in the one class is usually advantageous.

I'm somewhat uncertain whether an enumeration (CompilationLevel) is the most effective way to express this especially as all of this only seems relevant to CPAOT, not to the other CoreRT compilation modes. I wonder whether we could just have two instances of CompilationModuleGroup, one representing "inputs" and the other "bubble", and just patching its callers to use one or the other as appropriate.

Reply:
Another interesting question is whether the input assembly should have a level of its own (CompilationLevel.Input). This will become especially important once we start looking into single-file (merged) CPAOT build.

Reply:
`CanInline` API on `CompilationGroup` sounds like a good addition! `CompilationGroup` is a wrapper for policies and inlining is a policy.